### PR TITLE
Webassembly first pass

### PR DESCRIPTION
**Pull request by sirsquidness**
Make it able to build and run in webassembly 

The biggest problem this solves is that there's no local filesystem in webassembly, so I put in an interface for the caching layer so it could be removed for webassembly - the browser cache does all of the caching we need.

### CODE REVIEW

File: slippymap/cachedtileprovider.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+		// log.Print("Tile is cached:", tilePath)
+		return tilePath, nil
+	}
+}
+
+func newTransportWithLimitedConcurrency() http.RoundTripper {
+	t := http.DefaultTransport.(*http.Transport).Clone()
+
+	// We have 3 OSM CDN endpoints that we round robin, so max total conns is 3x this value
+	t.MaxConnsPerHost = 1
```
Comment (sirsquidness) [inline slippymap/cachedtileprovider.go:108]: I think I recall reading somewhere in a comment that we're meant to limit how hard we hit OSM's CDN yeah? Is that a thing, or am I imagining it? Either way, that's what I'm doing here,

File: main.go [hunk 2]
```diff
@@ -335,3 +325,34 @@ func main() {
 		log.Fatal(err)
 	}
 }
+
+// If we are running in WASM/JS, then the browser does all relevant tile caching for us.
+// If running in desktop app mode, we need to cache the tiles ourselves
+func tileProviderForOS() (slippymap.TileProvider, error) {
+	if runtime.GOOS == "js" || false {
```
Comment (mikenye) [inline main.go:332]: May I ask what the `|| false` is for?

File: slippymap/cachedtileprovider.go [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+	"os"
+	"path"
+)
+
+// TileProvider generates URLs (either https:// or file://) to the tile for the coord
+type TileProvider interface {
+	GetTileAddress(x, y, z int) (tilePath string, err error)
+}
+
+func NewCachedTileProvider(tileCachePath string, tileProvider TileProvider) *CachedTileProvider {
```
Comment (mikenye) [inline slippymap/cachedtileprovider.go:17]: I don't know what's going on here... :-)

Might need you to explain at some point.

File: slippymap/cachedtileprovider.go (hunk 1 above)
Comment (mikenye) [reply slippymap/cachedtileprovider.go:108]: Technical usage requirements ([here](https://operations.osmfoundation.org/policies/tiles/)) say:

> Maximum of 2 download threads. (Unmodified web browsers’ download thread limits are acceptable.)

**Review by mikenye — APPROVED** [review]

File: main.go (hunk 2 above)
Comment (sirsquidness) [reply main.go:332]: it is definitely not meant to be there and definitely does nothing 🙃 I think I was testing the build behaviour of `wasmserve` and just needed to force a rebuild by making a no-op change to the code.

File: slippymap/cachedtileprovider.go (hunk 3 above)
Comment (sirsquidness) [reply slippymap/cachedtileprovider.go:17]: Using interfaces so we can transparently swap out the caching vs not-caching implementation in wasm as necessary.

`TileProvider` -> something that generates tile addresses/URLs/whatever

`OSMTileProvider` -> generates the real OSM CDN addresses

`CachedTileProvider` -> generates on-disk URLs for cached files. It uses another TileProvider (OSMTileProvider) to generate the URLs to download in order to cache it.

Maybe the code would be clearer if I made the signature:

```
 func NewCachedTileProvider(tileCachePath string, tileProvider *OSMTileProvider) *CachedTileProvider 
```

File: main.go (hunk 2 above)
Comment (mikenye) [reply main.go:332]: Haha no worries. Will you update the branch in the other PR to remove it?

**Pull request merged**